In [141]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [142]:
train_data = pd.read_csv('../data/train.csv')
test_data = pd.read_csv('../data/test.csv')

In [143]:
train_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   str    
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   str    
 6   Alley          91 non-null     str    
 7   LotShape       1460 non-null   str    
 8   LandContour    1460 non-null   str    
 9   Utilities      1460 non-null   str    
 10  LotConfig      1460 non-null   str    
 11  LandSlope      1460 non-null   str    
 12  Neighborhood   1460 non-null   str    
 13  Condition1     1460 non-null   str    
 14  Condition2     1460 non-null   str    
 15  BldgType       1460 non-null   str    
 16  HouseStyle     1460 non-null   str    
 17  OverallQual    1460 non-null   int64  
 18  OverallCond    1460

In [144]:
train_data.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [145]:
missing_values = train_data.isnull().sum()
missing_values = missing_values[missing_values > 0]
missing_values.sort_values(ascending=False)

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
BsmtExposure      38
BsmtFinType2      38
BsmtQual          37
BsmtCond          37
BsmtFinType1      37
MasVnrArea         8
Electrical         1
dtype: int64

In [146]:
missing_percentage = (train_data.isnull().sum() / len(train_data)) * 100
missing_percentage = missing_percentage[missing_percentage > 0]
missing_percentage.sort_values(ascending=False)

PoolQC          99.520548
MiscFeature     96.301370
Alley           93.767123
Fence           80.753425
MasVnrType      59.726027
FireplaceQu     47.260274
LotFrontage     17.739726
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
BsmtExposure     2.602740
BsmtFinType2     2.602740
BsmtQual         2.534247
BsmtCond         2.534247
BsmtFinType1     2.534247
MasVnrArea       0.547945
Electrical       0.068493
dtype: float64

In [147]:
#filtering the rows where PoolQC is not null and selecting the columns PoolArea and PoolQC
train_data.loc[train_data["PoolQC"].notna(), ["PoolArea", "PoolQC"]]

#we now need to find out how many rows have PoolArea greater than 0 but PoolQC is null


,PoolArea,PoolQC
197,512,Ex
810,648,Fa
1170,576,Gd
1182,555,Ex
1298,480,Gd
1386,519,Fa
1423,738,Gd


In [148]:
train_data['PoolQC']=train_data['PoolQC'].fillna('NoPool')
train_data["PoolQC"].value_counts(dropna=False)

PoolQC
NoPool    1453
Gd           3
Ex           2
Fa           2
Name: count, dtype: int64

In [149]:
train_data["PoolArea"].value_counts().head(10)

PoolArea
0      1453
512       1
648       1
576       1
555       1
480       1
519       1
738       1
Name: count, dtype: int64

checking alley data to confirm that the nan is simply saying no alley so i will replace the empty cell with none

In [150]:
train_data["Alley"] = train_data['Alley'].fillna('NoAlley')
train_data["Alley"].value_counts(dropna=False)

Alley
NoAlley    1369
Grvl         50
Pave         41
Name: count, dtype: int64

comparing missing data from both masvnrarea and masvnrtype,  (masvnrtype)nan = 872 - 861 + 8 (0.0+nan)(masvnrArea)...  

In [151]:
train_data['MasVnrArea'] = train_data['MasVnrArea'].fillna(0.0)
train_data["MasVnrArea"].value_counts(dropna=False)

MasVnrArea
0.0      869
180.0      8
108.0      8
72.0       8
120.0      7
        ... 
426.0      1
96.0       1
438.0      1
194.0      1
119.0      1
Name: count, Length: 327, dtype: int64

In [152]:
train_data['MasVnrType'] = train_data['MasVnrType'].fillna('NoMasonry')
train_data["MasVnrType"].value_counts(dropna=False)

MasVnrType
NoMasonry    872
BrkFace      445
Stone        128
BrkCmn        15
Name: count, dtype: int64

this are the real missing values for MasVnrArea and MasVnrType, we can see that the MasVnrType is also null for these rows, so we can fill the MasVnrArea with 0 and MasVnrType with None..

In [153]:
train_data.loc[train_data["MasVnrArea"].isnull(), ["MasVnrArea", "MasVnrType"]]



,MasVnrArea,MasVnrType


In [154]:
train_data['BsmtQual'] = train_data['BsmtQual'].fillna("NoBasement")
train_data["BsmtQual"].value_counts(dropna=False)

BsmtQual
TA            649
Gd            618
Ex            121
NoBasement     37
Fa             35
Name: count, dtype: int64

with this information going with a fact that tis houses without the basement information smiply dosent have a basement so we can filll this empty cells with none or no basement

In [155]:
target_colunm = ['BsmtExposure','BsmtFinType1','BsmtFinType2','BsmtCond']
train_data[target_colunm] = train_data[target_colunm].fillna("NoBasement")

you see this colunm 948 it has a basement unlike the other having only a missing value so what i will do is to use MODE because its a type data so ill use most frequent data and impute it manually which is NO.

In [156]:
train_data.iloc[948, train_data.columns.get_loc("BsmtExposure")]

'NoBasement'

In [157]:
train_data["BsmtExposure"].value_counts(dropna=False)

BsmtExposure
No            953
Av            221
Gd            134
Mn            114
NoBasement     38
Name: count, dtype: int64

after i fill this cell will i fill the other nan with 'none' it has to be done in this step to ensure sucess 

In [158]:
#train_data["948,BsmtExposure"] = 'No'

what i see is a missing value in electrical this row has proen it self to hold alot of valuable information to remove it from the data set so we are going to fillna with the most frequent value 'SBrkr'

In [159]:
train_data['Electrical'].value_counts(dropna=False)

Electrical
SBrkr    1334
FuseA      94
FuseF      27
FuseP       3
Mix         1
NaN         1
Name: count, dtype: int64

In [160]:
train_data.iloc[1379]

Id                 1380
MSSubClass           80
MSZoning             RL
LotFrontage        73.0
LotArea            9735
                  ...  
MoSold                5
YrSold             2008
SaleType             WD
SaleCondition    Normal
SalePrice        167500
Name: 1379, Length: 81, dtype: object

In [161]:
train_data.loc[train_data["GarageType"].isnull(), ["GarageType", "GarageYrBlt", "GarageFinish", "GarageCars", "GarageArea", "GarageQual", "GarageCond"]]

,GarageType,GarageYrBlt,GarageFinish,GarageCars,GarageArea,GarageQual,GarageCond
39,NaN,NaN,NaN,0,0,NaN,NaN
48,NaN,NaN,NaN,0,0,NaN,NaN
78,NaN,NaN,NaN,0,0,NaN,NaN
88,NaN,NaN,NaN,0,0,NaN,NaN
89,NaN,NaN,NaN,0,0,NaN,NaN
...,...,...,...,...,...,...,...
1349,NaN,NaN,NaN,0,0,NaN,NaN
1407,NaN,NaN,NaN,0,0,NaN,NaN
1449,NaN,NaN,NaN,0,0,NaN,NaN
1450,NaN,NaN,NaN,0,0,NaN,NaN


In [162]:
#this is a type int64 data we will need to fill them empty values with 0.0
train_data["GarageYrBlt"].value_counts(dropna=False)

GarageYrBlt
NaN       81
2005.0    65
2006.0    59
2004.0    53
2003.0    50
          ..
1900.0     1
1927.0     1
1906.0     1
1908.0     1
1933.0     1
Name: count, Length: 98, dtype: int64

#A house has to be connected to the street, its impossible to fill the missing values with 0.0 because it will be misleading, we will fill them with the median value of the column because this is 50 percent of house in the area 69%

In [163]:

train_data["LotFrontage"].value_counts(dropna=False)

LotFrontage
NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
138.0      1
160.0      1
152.0      1
153.0      1
46.0       1
Name: count, Length: 111, dtype: int64

In [164]:
train_data["LotFrontage"].describe()

count    1201.000000
mean       70.049958
std        24.284752
min        21.000000
25%        59.000000
50%        69.000000
75%        80.000000
max       313.000000
Name: LotFrontage, dtype: float64

In [165]:
train_data.loc[train_data["LotFrontage"].isnull(), ["LotFrontage"]]

,LotFrontage
7,NaN
12,NaN
14,NaN
16,NaN
24,NaN
...,...
1429,NaN
1431,NaN
1441,NaN
1443,NaN


we will replace the NAN with NONE or no fence.  simply meanse majority of the houses in the area dont have fence

In [166]:
train_data["Fence"].value_counts(dropna=False)

Fence
NaN      1179
MnPrv     157
GdPrv      59
GdWo       54
MnWw       11
Name: count, dtype: int64

this will we be replaced with none because it the dataset and discription shows us majority of the house do not have extra features

In [167]:
train_data["MiscFeature"].value_counts(dropna=False)

MiscFeature
NaN     1406
Shed      49
Gar2       2
Othr       2
TenC       1
Name: count, dtype: int64

here we clean all the data

In [168]:
clean_data = train_data.copy()

In [169]:
#none_columns = ['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'MasVnrType']
#clean_data[none_columns] = clean_data[none_columns].fillna('None')

In [170]:
target_columns = [
    "BsmtQual",
    "BsmtCond",
    "BsmtExposure",
    "BsmtFinType1",
    "BsmtFinType2"
]

clean_data[target_columns] = clean_data[target_columns].fillna("NoBasement")

target_columns = [
    "GarageType",
    "GarageFinish",
    "GarageQual",
    "GarageCond"
]

clean_data[target_columns] = clean_data[target_columns].fillna("NoGarage")

garage_numeric = [
    "GarageYrBlt",
    "GarageCars",
    "GarageArea"
]

clean_data[garage_numeric] = clean_data[garage_numeric].fillna(0)

clean_data["Fence"] = clean_data["Fence"].fillna("NoFence")
clean_data["FireplaceQu"] = clean_data["FireplaceQu"].fillna("NoFireplace")
clean_data["Alley"] = clean_data["Alley"].fillna("NoAlley")
clean_data["MasVnrArea"] = clean_data["MasVnrArea"].fillna(0)
clean_data["MasVnrType"] = clean_data["MasVnrType"].fillna("NoFeature")
clean_data["MiscFeature"] = clean_data["MiscFeature"].fillna("NoMansonry")
clean_data["LotFrontage"] = clean_data.groupby("Neighborhood")["LotFrontage"].transform(
    lambda x: x.fillna(x.median())
)
clean_data["Electrical"] = clean_data["Electrical"].fillna(
    clean_data["Electrical"].mode()[0]
)
clean_data.isnull().sum().sort_values(ascending=False)

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
SalePrice        0
Length: 81, dtype: int64

In [171]:
#clean_data.loc[948, "BsmtExposure"] = 'No'
#clean_data.loc[1379, "Electrical"] = 'SBrkr'
#clean_data['MasVnrArea'] = clean_data['MasVnrArea'].fillna(0.0)
#clean_data['GarageYrBlt'] = clean_data['GarageYrBlt'].fillna(0.0)
#clean_data['LotFrontage'] = clean_data['LotFrontage'].fillna(clean_data['LotFrontage'].mean())
#clean_data.isnull().sum().sort_values(ascending=False)


In [172]:
clean_data.to_csv('../data/clean_house_price.csv',index=False)

In [173]:
clean_data['MiscFeature'].value_counts(dropna=False)

MiscFeature
NoMansonry    1406
Shed            49
Gar2             2
Othr             2
TenC             1
Name: count, dtype: int64

In [174]:
clean_data.to_csv("../data/clean_house_price.csv", index=False)

check = pd.read_csv("../data/clean_house_price.csv")

print(check["MiscFeature"].isnull().sum())
print(check["BsmtQual"].isnull().sum())

0
0


FEATURE ENCODING
the reason for creating a copy is to see how my model will work with and without the added features


In [175]:
clean_data['BsmtQual'].value_counts(dropna=False)

BsmtQual
TA            649
Gd            618
Ex            121
NoBasement     37
Fa             35
Name: count, dtype: int64